### PR TITLE
[widget] Improved widget/iframe communications

### PR DESCRIPTION
**Pull request by anxolin**
# Summary
This PR introduces 6 features or improvements to the widget communiations

## Motivation (context)
This PR is initially was born out a frustration I had with the widget messages. It was hard to know:
1. What is the list of messages the widget emits
2. What is the list of messages the widget listens
3. What are are the payloads for each message
4. Where in the code you emit, and where you consume

This lead to the first feature of this PR:

## Feature 1) Model widget events as an enum
Before the PR it would use loose strings. Now if we look for references for the enum, we can answer the question (4)

image

Also, we see improvement in (1) and (2) as at least we have a full list in the enum of events we EMIT or LISTEN.

## Feature 2) Split events the widget EMIT and the ones it LISTENs
The i created these 2 enums:

```ts
export enum WidgetMethodsEmit {
  ACTIVATE = 'ACTIVATE',
  UPDATE_HEIGHT = 'iframeHeight',
   ...
}

export enum WidgetMethodsListen {
  UPDATE_PARAMS = 'UPDATE_PARAMS',
  ...
}
```

This allows to know exactly what to expect from the widget.

## Feature 3) Define the payloads for all messages

This PR creates two payload mappings: `WidgetMethodsEmitPayloadMap`, `WidgetMethodsListenPayloadMap`

Also, all the associated types to make it work.

Now its easy to know what to expect. For example, the `UPDATE_HEIGHT` param will provide the `height` as an optional property containing a number:

image

## Feature 4) Refactor listeners and emits
The app would post messages and listen to messages in multiple places, and they were all repeating some logic, like the part of check the message is relative to the widget:

```ts
if (event.data.key !== COW_SWAP_WIDGET_EVENT_KEY || event.data.method !== method) {
      return
    }
```

Now there's a `listenToMessageFromWindow` which some fancy overload function, so you can use it for both, emit and listen events (so it can be used for bidirectional communication in and out of the widget)

## Feature 5) RPC provider messages
RPC provider messages were using a strange communication message:
- The RPC messages were being posted as they are in the window. So the only way to know is a RPC request or response is to check for the presence of a `jsonrpc` property in the message. This is fragile logic
- It was hard to find where the message is emited and received. The emit was hidden in a library we used `@ethvault/iframe-provider`, the reception was in the widget library, which it would post the response to the iframe also in the same strange format. Debugging this was a bit 🤯 for my last Friday

What i included is the declarative events they emit and receive, so we can benefit from features 1-4.

```ts
export enum WidgetMethodsEmit {
  PROVIDER_RPC_REQUEST = 'PROVIDER_RPC_REQUEST',
  ...
}

export enum WidgetMethodsListen {
  PROVIDER_RPC_RESPONSE = 'PROVIDER_RPC_RESPONSE',
 ... 
}
```

## Feature 6) Handle eip-1193 on messages
Same as  with the RPC messages, when we have a provider from the host app, we want to know when some events happen (like when you change chain). 

Before we had a subscription to the https://eips.ethereum.org/EIPS/eip-1193 `.on` events that would convert the message into a RPC call format (note that events are not a RPC messages) and posted into the window, again, relaying on this weird implementation from `@ethvault/iframe-provider`

For this I added a new event:

```ts
export enum WidgetMethodsListen {
  PROVIDER_ON_EVENT = 'PROVIDER_ON_EVENT',
 ... 
}
``` 

And turned the `IframeProvider` into a `WidgetEthereumProvider`. The name has widget now, as it is widget aware:
- Emits:  `PROVIDER_ON_EVENT` 
- Listens to: `PROVIDER_RPC_REQUEST`, `PROVIDER_RPC_RESPONSE`

# Not included
This PR still shows some issues with duplicated events. As this PR is already too big, I will do a follow up PR.

### CODE REVIEW

File: apps/cowswap-frontend/src/modules/injectedWidget/updaters/CowEventsUpdater.ts [hunk 1]
```diff
@@ -34,14 +28,10 @@ export function CowEventsUpdater() {
...
-    {
-      key: COW_SWAP_WIDGET_EVENT_KEY,
-      method: 'event',
-      eventName: event,
-      payload,
-    },
-    TARGET_ORIGIN
-  )
+function forwardEventToIframe<T extends CowEvents>(event: CowEvents, payload: CowEventPayloadMap[T]) {
+  // console.debug('[injectedWidget:forwardEventToIframe] Forward event', event, payload)
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/modules/injectedWidget/updaters/CowEventsUpdater.ts:32]: dead code?

File: libs/wallet/src/web3-react/connectors/Injected/WidgetEthereumProvider.ts [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+
+  // The event target. By default we use the window parent. This can be mocked for tests, or it can wrap
+  // a different interface, e.g. workers.
+  eventTarget?: Window
+}
+
+/**
+ * This is what we store in the state to keep track of pending promises.
+ */
+interface PromiseCompleter<TResult, TErrorData> {
```
Comment (alfetopito) [inline libs/wallet/src/web3-react/connectors/Injected/WidgetEthereumProvider.ts:55]: Nitpick reaaaaly minor: don't prefix types with `T`. It's already implicit.

File: libs/wallet/src/web3-react/connectors/Injected/WidgetEthereumProvider.ts [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+    this.reason = reason
+  }
+}
+
+/**
+ * This is the primary artifact of this library.
+ */
+export class WidgetEthereumProvider extends EventEmitter<IFrameEthereumProviderEventTypes> {
+  request({ method, params }: { method: string; params: unknown[] }) {
+    // console.log('[IFrameEthereumProvider] IFrameEthereumProvider - rpc request', { method, params })
```
Comment (alfetopito) [inline libs/wallet/src/web3-react/connectors/Injected/WidgetEthereumProvider.ts:125]: Keep or remove?

File: libs/wallet/src/web3-react/connectors/Injected/WidgetEthereumProvider.ts [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+    params: unknown[]
+  ): Promise<JsonRpcSucessfulResponseMessage<TResult> | JsonRpcErrorResponseMessage<TErrorData>> {
+    const id = getUniqueId()
+
+    const rpcRequest: JsonRpcRequestMessage = {
+      jsonrpc: JSON_RPC_VERSION,
+      id,
+      method,
+      params,
+      // ...(typeof params === 'undefined' ? null : { params }),
```
Comment (alfetopito) [inline libs/wallet/src/web3-react/connectors/Injected/WidgetEthereumProvider.ts:194]: Meant to be here for documentation?

File: libs/wallet/src/web3-react/connectors/Injected/WidgetEthereumProvider.ts [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+
+  /**
+   * Handle a Rpc Request
+   */
+  private handleRpcRequests = ({ rpcResponse }: ProviderRpcResponsePayload) => {
+    if (rpcResponse.id === undefined || rpcResponse.id === null) {
+      return
+    }
+
+    const completer = this.completers['' + rpcResponse.id]
```
Comment (alfetopito) [inline libs/wallet/src/web3-react/connectors/Injected/WidgetEthereumProvider.ts:277]: Why add it to an empty string?
Shorthand for converting the id itself to string?
I suggest less black magic and call id.toString() or String(id)

File: libs/widget-lib/src/IframeRpcProviderBridge.ts [hunk 6]
```diff
@@ -74,53 +81,53 @@ export class IframeRpcProviderBridge {
...
 
-    // Delegate the rpc call
-    this.processRequest(event.data)
+    this.processRpcRequest(rpcRequest)
   }
 
-  private onRpcEventForwardToIframe(event: string, params: unknown): void {
-    this.forwardRpcCallToIframe({ method: event, params: [params] })
+  private onProviderEvent(event: string, params: unknown): void {
+    // console.debug('[IframeRpcProviderBridge] onProviderEvent', { event, params })
```
Comment (alfetopito) [inline libs/widget-lib/src/IframeRpcProviderBridge.ts:119]: More debug logs. I'll stop mentioning them

File: libs/widget-lib/src/types.ts [hunk 7]
```diff
@@ -1,13 +1,30 @@
...
+}
+
 export interface JsonRpcRequest {
   id: number
   method: string
   params: unknown[]
 }
 
-// https://eips.ethereum.org/EIPS/eip-1193
+// https://eips.ethereum.org/EIPS/ei  p-1193
```
Comment (alfetopito) [inline libs/widget-lib/src/types.ts:27]: Ops, probably accident

**Review by alfetopito — COMMENTED** [review]
Mostly nitpicks, I'm not full in the loop to quickly understand the changes.
I'll delegate to Sasha's expertise.

File: libs/wallet/src/web3-react/connectors/Injected/WidgetEthereumProvider.ts (hunk 2 above)
Comment (anxolin) [reply libs/wallet/src/web3-react/connectors/Injected/WidgetEthereumProvider.ts]: they were copied from the library. Renamed to T and D (as its more usually for generics)

File: libs/wallet/src/web3-react/connectors/Injected/WidgetEthereumProvider.ts (hunk 3 above)
Comment (anxolin) [reply libs/wallet/src/web3-react/connectors/Injected/WidgetEthereumProvider.ts:125]: Ok, I will remove but it was like: i find it nice for debugging in the future, but I don't necesarilly want to log it unless I need to debug the RPC calls to the provider outside of the widget.

File: libs/wallet/src/web3-react/connectors/Injected/WidgetEthereumProvider.ts (hunk 4 above)
Comment (anxolin) [reply libs/wallet/src/web3-react/connectors/Injected/WidgetEthereumProvider.ts:194]: nup, deleting

File: libs/wallet/src/web3-react/connectors/Injected/WidgetEthereumProvider.ts (hunk 5 above)
Comment (anxolin) [reply libs/wallet/src/web3-react/connectors/Injected/WidgetEthereumProvider.ts:277]: not my code. I think is how they converted to string the number...

I prefer to touch the minimum of this library

File: apps/cowswap-frontend/src/modules/injectedWidget/updaters/CowEventsUpdater.ts (hunk 1 above)
Comment (anxolin) [reply apps/cowswap-frontend/src/modules/injectedWidget/updaters/CowEventsUpdater.ts:32]: https://github.com/cowprotocol/cowswap/pull/3911

File: libs/wallet/src/web3-react/connectors/Injected/WidgetEthereumProvider.ts (hunk 3 above)
Comment (anxolin) [reply libs/wallet/src/web3-react/connectors/Injected/WidgetEthereumProvider.ts]: Deleted this and all of the other dead codes for console messages (from this PR and others) https://github.com/cowprotocol/cowswap/pull/3911

File: libs/widget-lib/src/IframeRpcProviderBridge.ts (hunk 6 above)
Comment (anxolin) [reply libs/widget-lib/src/IframeRpcProviderBridge.ts:119]: Deleted this and all of the other dead codes for console messages (from this PR and others) https://github.com/cowprotocol/cowswap/pull/3911

File: libs/widget-lib/src/types.ts (hunk 7 above)
Comment (anxolin) [reply libs/widget-lib/src/types.ts:27]: f140d14f58ea2941f9f8faab2fb0eb9065501a69

**Comment by anxolin** [comment]
Merging, i address any other comment after

File: libs/wallet/src/web3-react/connectors/Injected/WidgetEthereumProvider.ts (hunk 2 above)
Comment (alfetopito) [reply libs/wallet/src/web3-react/connectors/Injected/WidgetEthereumProvider.ts]: Ah ok. If this was copied over the definitely ignore.

**Pull request merged**